last batch of .NET guideline gap cleanup (#2361)
Azure/azure-sdk: docs/dotnet/introduction.md

}
```

> The _Options_ class is designed similarly to .NET custom attributes, where required service method parameters are modeled as _Options_ class constructor parameters and get-only properties, and optional parameters are get-set properties. 

Reviewer: > The _Options_ class is designed similarly to .NET custom attributes,

😲😲

I had no idea!  Thank you for adding this!!

> The _Options_ class is designed similarly to .NET custom attributes, where required service method parameters are modeled as _Options_ class constructor parameters and get-only properties, and optional parameters are get-set properties. 

Reviewer 2: The way it's written makes me think there's a class called just `Options`.  What about something like:
```suggestion
These _Options_ classes are designed similarly to .NET custom attributes:
- Required service method parameters are modeled as _Options_ class constructor parameters and get-only properties.
- Optional service method parameters are modeled as get-set properties. 
```

{% include requirement/MUST id="dotnet-params-complex" %} use the _options_ parameter pattern for complex service methods.

> {% include requirement/MAY id="dotnet-params-complex" %} use the _options_ parameter pattern for simple service methods that you expect to `grow` in the future.

Reviewer: Below:
> {% include requirement/MAY id="dotnet-params-complex" %} add simple overloads of methods using the _options_ parameter pattern.

It's slightly unclear from the way this is worded whether you can add simple overloads that use the _options_ pattern, or whether you can add simple overloads that take the commonly used parameters directly, as overloads of methods using the _options_ pattern.  Would you be willing to shift the language slightly to make the intent clear?

Reviewer: Oh, actually, I guess the statement below clarifies this, but if there's a way to highlight that more, I think it would be helpful to the reader.

Author: I added a code sample. I think it will clarify.

If in common scenarios, users are likely to pass just a small subset of what the _options_ parameter represents, consider adding an overload with a parameter list representing just this subset.

```csharp
    // main overload taking the options property bag

>     public virtual Response<BlobInfo> CreateBlob(BlobCreateOptions options = null, CancellationToken cancellationToken = default);

Reviewer 2: Do you think we should add the `Async` signatures to make it clear that you need to provide every overload for both?  We have separate guidelines, but it might help newer folks make that connection.

    // simple overload with a subset of parameters of the options bag

>     public virtual Response<BlobContainerInfo> CreateBlob(string blobName, CancellationToken cancellationToken = default);

Reviewer 2: _(I'm commenting on this here since Github won't let me comment on the example above since it wasn't touched in this PR)_

Your example `BlobCreateOptions` above doesn't have a `BlobName` property so it doesn't really make sense to have some overloads take `blobName` and some only take `options`.  I don't think the existing sample follows the guidelines yet.

##### Model Type Naming

> TODO: issue #2298

Reviewer 2: ```suggestion
TODO: https://github.com/azure/azure-sdk/issues/2298
```

#### Common Libraries

There are occasions when common code needs to be shared between several client libraries. For example, a set of cooperating client libraries may wish to share a set of exceptions or models.

> {% include requirement/MUST id="dotnet-commonlib-approval" %} gain [Architecture Board] approval prior to implementing a common library.

Reviewer: 👍

Anything worth calling out that API designers can pre-consider before approaching arch board?  Stuff you wish they had thought about before coming to you?  Ways you'll make trade-offs around this?  Common pitfalls, etc?

(if not, no worries)

> [Architecture Board] 
Provide a link for this?

Author: I think this is out off scope of this document. Shared would be a document (or book) on its own.

#### Common Libraries

> There are occasions when common code needs to be shared between several client libraries. For example, a set of cooperating client libraries may wish to share a set of exceptions or models.

Reviewer 2: Should we add a guideline saying something like "CONSIDER using internal shared source for common code when possible"?

> {% include requirement/MUST id="dotnet-commonlib-approval" %} gain [Architecture Board] discuss how to design such common library.

Reviewer 2: Feels like something's missing here.  Maybe:

```suggestion
{% include requirement/MUST id="dotnet-commonlib-approval" %} gain [Architecture Board approval](https://azure.github.io/azure-sdk/policies_reviewprocess.html) to design a common library.
```

### Documentation Style

> TODO: issue #2338

Reviewer 2: ```suggestion
TODO: https://github.com/azure/azure-sdk/issues/2338
```
